parachain/primitives/testutils/src/lib.rs

```diff
@@ -0,0 +1,129 @@
+use ethereum_types::{H128, H256, H512};
+use serde::{Deserialize, Deserializer};
+
+/// The structs defined below are used to load Ethash merkle proofs
+/// generated by https://github.com/talbaneth/ethashproof.
+/// To generate proof JSON:
+///     ./cmd/relayer/relayer $BLOCK_NUM
+/// To load in test:
+///     $proof_data = BlockWithProofs::from_file("path/to/proof.json")
+///         .to_double_node_with_merkle_proof_vec();
+
+#[derive(Clone)]
+pub struct Hex(pub Vec<u8>);
+
+impl<'de> Deserialize<'de> for Hex {
+    fn deserialize<D>(deserializer: D) -> Result<Self, <D as Deserializer<'de>>::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let mut s = <String as Deserialize>::deserialize(deserializer)?;
+        if s.starts_with("0x") {
+            s = s[2..].to_string();
+        }
+        if s.len() % 2 == 1 {
+            s.insert_str(0, "0");
+        }
+        let v: Vec<u8> = hex::FromHexIter::new(&s)
+            .map(|x| x.unwrap())
+            .collect();
+        Ok(Hex(v))
+    }
+}
+
+impl From<&Hex> for H256 {
+    fn from(item: &Hex) -> Self {
+        let mut data = [0u8; 32];
+        let size = item.0.len();
+        for i in 0..size {
+            data[31 - i] = item.0[size - 1 - i];
+        }
+        data.into()
+    }
+} 
+
+impl From<&Hex> for H128 {
+    fn from(item: &Hex) -> Self {
+        let mut data = [0u8; 16];
+        let size = item.0.len();
+        for i in 0..size {
+            data[15 - i] = item.0[size - 1 - i];
+        }
+        data.into()
+    }
+} 
+
+#[derive(Deserialize)]
+struct BlockWithProofsRaw {
+    pub proof_length: u64,
+    pub header_rlp: Hex,
+    pub merkle_root: Hex,        // H128
+    pub elements: Vec<Hex>,      // H256
+    pub merkle_proofs: Vec<Hex>, // H128
+}
+
+pub struct BlockWithProofs {
+    pub proof_length: u64,
+    pub header_rlp: Hex,
+    pub merkle_root: H128,
+    pub elements: Vec<H256>,
+    pub merkle_proofs: Vec<H128>,
+}
+
+impl From<BlockWithProofsRaw> for BlockWithProofs {
+    fn from(item: BlockWithProofsRaw) -> Self {
+        Self {
+            proof_length: item.proof_length,
+            header_rlp: item.header_rlp,
+            merkle_root: (&item.merkle_root).into(),
+            elements: item.elements.iter().map(|e| e.into()).collect(),
+            merkle_proofs: item
+                .merkle_proofs
+                .iter()
+                .map(|e| e.into())
+                .collect(),
+        }
+    }
+}
+
+impl BlockWithProofs {
+    pub fn from_file(filename: &str) -> Self {
+        let raw: BlockWithProofsRaw =
+            serde_json::from_reader(std::fs::File::open(std::path::Path::new(filename)).unwrap()).unwrap();
```

Comment:
No need to use `Path::new` here since `&str` implements `AsRef<Path>`: https://doc.rust-lang.org/std/convert/trait.AsRef.html#impl-AsRef%3CPath%3E

```suggestion
            serde_json::from_reader(std::fs::File::open(filename).unwrap()).unwrap();
```